File: src/jvm/jnet/src/main/java/org/mases/jnet/developed/JNetHelper.java
/*
 *  Copyright 2025 MASES s.r.l.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *  http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 *
 *  Refer to LICENSE for more information.
 */

package org.mases.jnet.developed;

import java.nio.ByteBuffer;
import java.nio.CharBuffer;
import java.nio.DoubleBuffer;
import java.nio.FloatBuffer;
import java.nio.IntBuffer;
import java.nio.LongBuffer;
import java.nio.ShortBuffer;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

public class JNetHelper {
	public static List<?> listFromPrimitiveArray(Object input) {
		if (input == null) throw new IllegalArgumentException("Input parameter is null");
		
		if (input instanceof boolean[]) {
			final List<Boolean> l = new ArrayList<Boolean>();
			for (final boolean s : (boolean[])input) {
				l.add(s);
			}
			return l;
		}
		else if (input instanceof byte[]) {
			final List<Byte> l = new ArrayList<Byte>();
			for (final byte s : (byte[])input) {
				l.add(s);
			}
			return l;
		}
		else if (input instanceof char[]) {
			final List<Character> l = new ArrayList<Character>();
			for (final char s : (char[])input) {
				l.add(s);
			}
			return l;
		}
		else if (input instanceof short[]) {
			final List<Short> l = new ArrayList<Short>();
			for (final short s : (short[])input) {
				l.add(s);
			}
			return l;
		}
		else if (input instanceof int[]) {
			Integer[] array = Arrays.stream((int[])input).boxed().toArray(Integer[]::new);
			return Arrays.asList(array);
		}
		else if (input instanceof long[]) {
			Long[] array = Arrays.stream((long[])input).boxed().toArray(Long[]::new);
			return Arrays.asList(array);
		}
		else if (input instanceof float[]) {
			final List<Float> l = new ArrayList<Float>();
			for (final float s : (float[])input) {
				l.add(s);
			}
			return l;
		}
		else if (input instanceof double[]) {
			Double[] array = Arrays.stream((double[])input).boxed().toArray(Double[]::new);
			return Arrays.asList(array);
		}

		if(!input.getClass().isArray()) throw new IllegalArgumentException("Input parameter is not an array");

		throw new ClassCastException(input.getClass().getName() + " cannot be converted with this function");
	}

	public static List<Byte> listFromByteBuffer(ByteBuffer input) {
		if (input == null) throw new IllegalArgumentException("Input parameter is null");
		
		input.rewind();
		int length = input.remaining();
		final ArrayList<Byte> l = new ArrayList<Byte>(length);
		for (int i = 0; i < length; i++) {
			l.set(i, Byte.valueOf(input.get(i)));
		}
		return l;
	}
	
	public static List<Character> listFromCharBuffer(CharBuffer input) {
		if (input == null) throw new IllegalArgumentException("Input parameter is null");

		input.rewind();
		int length = input.remaining();
		final ArrayList<Character> l = new ArrayList<Character>(length);
		for (int i = 0; i < length; i++) {
			l.set(i, Character.valueOf(input.get(i)));
		}
		return l;
	}
	
	public static List<Double> listFromDoubleBuffer(DoubleBuffer input) {
		if (input == null) throw new IllegalArgumentException("Input parameter is null");

		input.rewind();
		int length = input.remaining();
		final ArrayList<Double> l = new ArrayList<Double>(length);
		for (int i = 0; i < length; i++) {
			l.set(i, Double.valueOf(input.get(i)));
		}
		return l;
	}
	
	public static List<Float> listFromFloatBuffer(FloatBuffer input) {
		if (input == null) throw new IllegalArgumentException("Input parameter is null");

		input.rewind();
		int length = input.remaining();
		final ArrayList<Float> l = new ArrayList<Float>(length);
		for (int i = 0; i < length; i++) {
			l.set(i, Float.valueOf(input.get(i)));
		}
		return l;
	}
	
	public static List<Integer> listFromIntBuffer(IntBuffer input) {
		if (input == null) throw new IllegalArgumentException("Input parameter is null");

		input.rewind();
		int length = input.remaining();
		final ArrayList<Integer> l = new ArrayList<Integer>(length);
		for (int i = 0; i < length; i++) {
			l.set(i, Integer.valueOf(input.get(i)));
		}
		return l;
	}
	
	public static List<Long> listFromLongBuffer(LongBuffer input) {
		if (input == null) throw new IllegalArgumentException("Input parameter is null");

		input.rewind();
		int length = input.remaining();
		final ArrayList<Long> l = new ArrayList<Long>(length);
		for (int i = 0; i < length; i++) {
			l.set(i, Long.valueOf(input.get(i)));
		}
		return l;
	}
	
	public static List<Short> listFromShortBuffer(ShortBuffer input) {
		if (input == null) throw new IllegalArgumentException("Input parameter is null");

		input.rewind();
		int length = input.remaining();
		final ArrayList<Short> l = new ArrayList<Short>(length);
		for (int i = 0; i < length; i++) {
			l.set(i, Short.valueOf(input.get(i)));
		}
		return l;
	}
}

File: src/net/JNet/Specific/JNetHelper.cs
/*
*  Copyright (c) 2022-2025 MASES s.r.l.
*
*  Licensed under the Apache License, Version 2.0 (the "License");
*  you may not use this file except in compliance with the License.
*  You may obtain a copy of the License at
*
*  http://www.apache.org/licenses/LICENSE-2.0
*
*  Unless required by applicable law or agreed to in writing, software
*  distributed under the License is distributed on an "AS IS" BASIS,
*  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
*  See the License for the specific language governing permissions and
*  limitations under the License.
*
*  Refer to LICENSE for more information.
*/

using Java.Nio;
using Java.Util;
using MASES.JCOBridge.C2JBridge;
using MASES.JCOBridge.C2JBridge.JVMInterop;
using System;

namespace MASES.JNet.Specific
{
    /// <summary>
    /// Helper class of JNet for some specific operations
    /// </summary>
    public sealed class JNetHelper : MASES.JCOBridge.C2JBridge.JVMBridgeBase<JNetHelper>
    {
        const string _bridgeClassName = "org.mases.jnet.developed.JNetHelper";

        /// <inheritdoc/>
        public JNetHelper() { }
        /// <inheritdoc/>
        public JNetHelper(IJVMBridgeBaseInitializer initializer) : base(initializer) { }

        private static readonly global::System.Exception _LocalBridgeClazzException = null;
        private static readonly MASES.JCOBridge.C2JBridge.JVMInterop.IJavaType _LocalBridgeClazz = JVMBridgeBase.ClazzOf(_bridgeClassName, out _LocalBridgeClazzException, false);
        private static MASES.JCOBridge.C2JBridge.JVMInterop.IJavaType LocalBridgeClazz => _LocalBridgeClazz ?? throw _LocalBridgeClazzException ?? new global::System.InvalidOperationException($"Class {_bridgeClassName} was not found.");

        /// <inheritdoc/>
        public override string BridgeClassName => _bridgeClassName;
        /// <inheritdoc/>
        public override bool IsBridgeAbstract => false;
        /// <inheritdoc/>
        public override bool IsBridgeCloseable => false;
        /// <inheritdoc/>
        public override bool IsBridgeInterface => false;
        /// <inheritdoc/>
        public override bool IsBridgeStatic => false;

        /// <summary>
        /// Retrieve the <see cref="IJVMBridgeDefinition.BridgeClassName"/> of <typeparamref name="TClass"></typeparamref>
        /// </summary>
        /// <typeparam name="TClass">A type implementing <see cref="IJVMBridgeBase"/></typeparam>
        /// <returns>The <see cref="IJVMBridgeDefinition.BridgeClassName"/></returns>
        public static Java.Lang.Class<TClass> Class<TClass>() where TClass : IJVMBridgeBase, new()
        {
            var className = JVMBridgeBase.ClassNameOf<TClass>();
            return Java.Lang.Class<TClass>.SExecute<Java.Lang.Class<TClass>>("forName", className);
        }

        /// <summary>
        /// Executes the <see href="https://docs.oracle.com/en%2Fjava%2Fjavase%2F11%2Fdocs%2Fapi%2F%2F/java.base/java/util/Arrays.html#asList(T...)"/> over primitive types
        /// </summary>
        /// <param name="data">Array of primitive types</param>
        /// <param name="buffered">Set to <see langword="true"/> to use <see cref="ByteBuffer"/>, <see cref="IntBuffer"/>, <see cref="DoubleBuffer"/> and so on to transfer data to JVM</param>
        /// <returns>A <see cref="Java.Util.List"/> can be used as input of <see cref="ArrayList{T}"/></returns>
        public static List ListFrom<T>(T[] data, bool buffered = false)
        {
            if (data == null) throw new ArgumentNullException(nameof(data));
            if (!buffered)
            {
                return SExecute<List>(LocalBridgeClazz, "listFromPrimitiveArray", data);
            }
            else
            {
                var buf = JCOBridge.C2JBridge.JCOBridge.Global.JVM.NewDirectBuffer(data, false, false);
                ByteBuffer bb = JVMBridgeBase.WrapsDirect<ByteBuffer>(buf.DisableCleanupAndReturn());
                if (data is byte[]) return ListFrom(bb);
                else
                {
                    bb.Order(BitConverter.IsLittleEndian ? ByteOrder.LITTLE_ENDIAN : ByteOrder.BIG_ENDIAN);
                    if (data is char[]) return ListFrom(bb.AsCharBuffer());
                    else if (data is double[]) return ListFrom(bb.AsDoubleBuffer());
                    else if (data is float[]) return ListFrom(bb.AsFloatBuffer());
                    else if (data is int[]) return ListFrom(bb.AsIntBuffer());
                    else if (data is long[]) return ListFrom(bb.AsLongBuffer());
                    else if (data is short[]) return ListFrom(bb.AsShortBuffer());
                    else throw new System.InvalidCastException($"{typeof(T)} does not have a ready made counter part");
                }
            }
        }

        /// <summary>
        /// Executes the <see href="https://docs.oracle.com/en%2Fjava%2Fjavase%2F11%2Fdocs%2Fapi%2F%2F/java.base/java/util/Arrays.html#asList(T...)"/> using a <see cref="ByteBuffer"/> in <paramref name="arg0"/>
        /// </summary>
        /// <param name="arg0"><see cref="ByteBuffer"/> to be used</param>
        /// <returns>A <see cref="Java.Util.List"/> can be used as input of <see cref="ArrayList{T}"/></returns>
        public static List<Java.Lang.Byte> ListFrom(ByteBuffer arg0)
        {
            return SExecute<List<Java.Lang.Byte>>(LocalBridgeClazz, "listFromByteBuffer", arg0);
        }

        /// <summary>
        /// Executes the <see href="https://docs.oracle.com/en%2Fjava%2Fjavase%2F11%2Fdocs%2Fapi%2F%2F/java.base/java/util/Arrays.html#asList(T...)"/> using a <see cref="CharBuffer"/> in <paramref name="arg0"/>
        /// </summary>
        /// <param name="arg0"><see cref="CharBuffer"/> to be used</param>
        /// <returns>A <see cref="Java.Util.List"/> can be used as input of <see cref="ArrayList{T}"/></returns>
        public static List<Java.Lang.Character> ListFrom(CharBuffer arg0)
        {
            return SExecute<List<Java.Lang.Character>>(LocalBridgeClazz, "listFromCharBuffer", arg0);
        }

        /// <summary>
        /// Executes the <see href="https://docs.oracle.com/en%2Fjava%2Fjavase%2F11%2Fdocs%2Fapi%2F%2F/java.base/java/util/Arrays.html#asList(T...)"/> using a <see cref="DoubleBuffer"/> in <paramref name="arg0"/>
        /// </summary>
        /// <param name="arg0"><see cref="DoubleBuffer"/> to be used</param>
        /// <returns>A <see cref="Java.Util.List"/> can be used as input of <see cref="ArrayList{T}"/></returns>
        public static List<Java.Lang.Double> ListFrom(DoubleBuffer arg0)
        {
            return SExecute<List<Java.Lang.Double>>(LocalBridgeClazz, "listFromDoubleBuffer", arg0);
        }

        /// <summary>
        /// Executes the <see href="https://docs.oracle.com/en%2Fjava%2Fjavase%2F11%2Fdocs%2Fapi%2F%2F/java.base/java/util/Arrays.html#asList(T...)"/> using a <see cref="FloatBuffer"/> in <paramref name="arg0"/>
        /// </summary>
        /// <param name="arg0"><see cref="FloatBuffer"/> to be used</param>
        /// <returns>A <see cref="Java.Util.List"/> can be used as input of <see cref="ArrayList{T}"/></returns>
        public static List<Java.Lang.Float> ListFrom(FloatBuffer arg0)
        {
            return SExecute<List<Java.Lang.Float>>(LocalBridgeClazz, "listFromFloatBuffer", arg0);
        }

        /// <summary>
        /// Executes the <see href="https://docs.oracle.com/en%2Fjava%2Fjavase%2F11%2Fdocs%2Fapi%2F%2F/java.base/java/util/Arrays.html#asList(T...)"/> using a <see cref="IntBuffer"/> in <paramref name="arg0"/>
        /// </summary>
        /// <param name="arg0"><see cref="IntBuffer"/> to be used</param>
        /// <returns>A <see cref="Java.Util.List"/> can be used as input of <see cref="ArrayList{T}"/></returns>
        public static List<Java.Lang.Integer> ListFrom(IntBuffer arg0)
        {
            return SExecute<List<Java.Lang.Integer>>(LocalBridgeClazz, "listFromIntBuffer", arg0);
        }

        /// <summary>
        /// Executes the <see href="https://docs.oracle.com/en%2Fjava%2Fjavase%2F11%2Fdocs%2Fapi%2F%2F/java.base/java/util/Arrays.html#asList(T...)"/> using a <see cref="LongBuffer"/> in <paramref name="arg0"/>
        /// </summary>
        /// <param name="arg0"><see cref="LongBuffer"/> to be used</param>
        /// <returns>A <see cref="Java.Util.List"/> can be used as input of <see cref="ArrayList{T}"/></returns>
        public static List<Java.Lang.Long> ListFrom(LongBuffer arg0)
        {
            return SExecute<List<Java.Lang.Long>>(LocalBridgeClazz, "listFromLongBuffer", arg0);
        }

        /// <summary>
        /// Executes the <see href="https://docs.oracle.com/en%2Fjava%2Fjavase%2F11%2Fdocs%2Fapi%2F%2F/java.base/java/util/Arrays.html#asList(T...)"/> using a <see cref="ShortBuffer"/> in <paramref name="arg0"/>
        /// </summary>
        /// <param name="arg0"><see cref="ShortBuffer"/> to be used</param>
        /// <returns>A <see cref="Java.Util.List"/> can be used as input of <see cref="ArrayList{T}"/></returns>
        public static List<Java.Lang.Short> ListFrom(ShortBuffer arg0)
        {
            return SExecute<List<Java.Lang.Short>>(LocalBridgeClazz, "listFromShortBuffer", arg0);
        }
    }
}
